src/Universum/Container/Class.hs

```diff
@@ -286,19 +167,62 @@ instance ToPairs (Map k v) where
     elems   = M.elems
     {-# INLINE elems #-}
 
-
 ----------------------------------------------------------------------------
--- Additional operations that don't make much sense for e.g. Maybe
+-- Containers (e.g. tuples and Maybe aren't containers)
 ----------------------------------------------------------------------------
 
--- | A class for 'ToList's that aren't trivial like 'Maybe' (e.g. can hold
--- more than one value)
-class ToList t => Container t where
+-- | Default implementation of 'Element' associated type family.
+type family ElementDefault (t :: *) :: * where
+    ElementDefault (f a) = a
+
+-- | Very similar to 'Foldable' but also allows instances for monomorphic types
+-- like 'Text' but forbids instances for 'Maybe' and similar. This class is used as
+-- a replacement for 'Foldable' type class. It solves the following problems:
+--
+-- 1. 'length', 'foldr' and other functions work on more types for which it makes sense.
+-- 2. You can't accidentally use 'length' on polymorphic 'Foldable' (like list),
+--    replace list with 'Maybe' and then debug error for two days.
+-- 3. More efficient implementaions of functions for polymorphic types (like 'elem' for 'Set').
+--
+-- The drawbacks:
+--
+-- 1. Type signatures of polymorphic functions look more scary.
+-- 2. Orphan instances are involved if you want to use 'foldr' (and similar) on types from libraries.
+class Container t where
+    -- | Type of element for some container. Implemented as an asscociated type family because
+    -- some containers are monomorphic over element type (like 'T.Text', 'IntSet', etc.)
+    -- so we can't implement nice interface using old higher-kinded types
+    -- approach. Implementing this as an associated type family instead of
+    -- top-level family gives you more control over element types.
+    type Element t :: *
+    type Element t = ElementDefault t
+
     -- | Constraint for elements. This can be used to implement more efficient
-    -- implementation of some methods.
+    -- implementation of some methods. For example 'elem' for 'Set' and 'HashSet'.
     type ElementConstraint t :: * -> Constraint
     type ElementConstraint t = Eq
 
+    -- | Convert container to list of elements.
+    --
+    -- >>> toList @Text "aba"
+    -- "aba"
+    -- >>> :t toList @Text "aba"
+    -- toList @Text "aba" :: [Char]
+    toList :: t -> [Element t]
+    default toList :: (Foldable f, t ~ f a, Element t ~ a) => t -> [Element t]
+    toList = Foldable.toList
+    {-# INLINE toList #-}
+
+    -- | Checks whether container is empty.
+    --
+    -- >>> null @Text ""
+    -- True
+    -- >>> null @Text "aba"
+    -- False
+    null :: t -> Bool
+    null = List.null . toList
```

Comment:
I've noticed that there is `default` line for `toList`, but not for `null`

Reply:
Good point :+1: This can (and should) be done since `null` in `base` package is a member of `Foldable` type class